### PR TITLE
Optimize React DevTools unmount handling

### PR DESCRIPTION
**Pull request by markerikson**
This PR:

- Reworks the React DevTools stub logic to capture all unmounted React Fiber objects in a `Set`, which is then brought into scope in the next `onCommitFiberRoot` call.  This allows the routine to run the unmount handling for each fiber inside the `"commit"` eval step, thus batching up the processing and eliminating the need to evaluate the 80-90% of the points that were `"unmount"` annotations.

### CODE REVIEW

File: third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc [hunk 1]
```diff
@@ -2854,32 +2862,45 @@ let uidCounter = 0;
...
   
-  window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook", "commit-fiber-root");
+  window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook", JSON.stringify({type: "commit-fiber-root"}));
+
+  unmountedFibersSet.clear();
 }
 
 function onPostCommitFiberRoot(rendererID, root) {
-  window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook", "post-commit-fiber-root");
+  window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook", JSON.stringify({type: "post-commit-fiber-root"}));
```
Comment (loganfsmyth) [inline third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:2903]: Looks like we may not even need this change to have JSON then?

File: third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc [hunk 2]
```diff
@@ -2854,32 +2862,45 @@ let uidCounter = 0;
...
+  // Unmounts are always one fiber at a time during the commit phase.
+  // Stash the unmounted fibers here, so we can map them to persistent 
+  // object IDs inside of `onCommitFiberRoot` processing in the routine.
+  const unmountedFibersSet = getUnmountedFibersSetForRenderer(rendererID);
+  unmountedFibersSet.add(fiber);
+
+  window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook", JSON.stringify({type: "commit-fiber-unmount"}));
 }
 
 function onCommitFiberRoot(rendererID, root, priorityLevel) {
```
Comment (loganfsmyth) [inline third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:2880]: Not strictly related to this PR, but could we also add something like
```
const annotationType = "commit-fiber-root";
``` 
in each scope where we trigger the annotation too? It'd be great if the code running at the annotations didn't have to have all that hacky `new Error().stack` processing that I added to decide which function it is inside of.

Comment (loganfsmyth) [reply third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:2895]: I think either way we're dropping backward-compat with all this new stuff.

File: third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc [hunk 3]
```diff
@@ -2860,33 +2871,57 @@ Object.defineProperty(window, "__REACT_DEVTOOLS_GLOBAL_HOOK__", {
...
     mountedRoots.add(root);
   } else if (isKnownRoot && isUnmounting) {
     mountedRoots.delete(root);
   }
+
+  // Get these so it's in scope in the routine eval, and we can clear it after the annotation
+  const unmountedFibersSet = getUnmountedFibersSetForRenderer(rendererID);
   
-  window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook", "commit-fiber-root");
+  window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook", JSON.stringify({type: annotationType}));
```
Comment (loganfsmyth) [inline third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:2917]: How would you feel about having this do
```
window.__RECORD_REPLAY_ANNOTATION_HOOK__("react-devtools-hook:v1:" + annotationType, "");
```
to indicate some basic versioning info in the recordings, and to include the annotation type in the kind itself? That also gives us more explicit namespacing and moves more info into the `kind` and out of `contents`.

My main concern is that as soon as we embed this in JSON, we then have to either manually parse that JSON and query it, so we're limiting ourselves to having to fetch all the points, do some preprocessing, and then send a protocol command, and it seems like it'd be awesome to be able to do
```
runEvaluation({
  pointSelector: {
    kind: "annotations",
    types: [
      "react-devtools-hook:v1:commit-fiber-root", 
    ],
  }
})
```
to run an evaluation without having to fetch all the data to the client first.

Comment (markerikson) [reply third_party/blink/renderer/bindings/core/v8/record_replay_interface.cc:2917]: Sure, let's do that.  Gonna knock out a Passport Tour thing for Jon and then I'll make those changes.

**Review by loganfsmyth — APPROVED** [review]

**Pull request merged**